[#125] 'Baking with tezos-packaging on Ubuntu' article (#131)
serokell/tezos-packaging: docs/baking.md

<!--
   - SPDX-FileCopyrightText: 2019 TQ Tezos <https://tqtezos.com/>
   -
   - SPDX-License-Identifier: LicenseRef-MIT-TQ
   -->
# Baking with tezos-packaging on Ubuntu

Tezos-packaging provides an easy way to install and set up infrastructure for

> interacting with Tezos blockchain. This article provides a step-by-step guide for
> setting up baking instance for Tezos on Ubuntu.

Reviewer: ```suggestion
interacting with Tezos blockchain.

This article provides a step-by-step guide for setting up baking instance for Tezos on Ubuntu.
```
Just for legibility.

## Prerequisites

### Installing required packages

> In order to run baking instance, you'll need the following Tezos binaries:

Reviewer: ```suggestion
In order to run a baking instance, you'll need the following Tezos binaries:
```

Author: Articles have always been tricky for me :joy: :

Reviewer: Naaa, these are just minor fixes, overall it's well written IMO

`tezos-client`, `tezos-node`, `tezos-baker-<proto>`, `tezos-endorser-<proto>`.

To install them run the following commands:
```
# Add PPA with Tezos binaries
sudo add-apt-repository ppa:serokell/tezos
sudo apt-get update
# Install packages
sudo apt-get install tezos-client tezos-node tezos-baker-<proto> tezos-endorser-<proto>
```

Packages for `tezos-node`, `tezos-baker-<proto>` and `tezos-endorser-<proto>` provide

> systemd units for running the corresponding binaries in the background

Reviewer: ```suggestion
systemd units for running the corresponding binaries in the background.
```

> 

Reviewer: I think it's fairly clear what `<proto>` is, but still we should probably list the options here somehow IMO.

Especially because it seems that the names are not exactly the same as the one from our release page (although only for capitalization AFAICT).

What do you think about and adding a short sentence here pointing at launchpad.net?
E.g. for `baker`s link to:
https://launchpad.net/~serokell/+archive/ubuntu/tezos/+packages?field.name_filter=tezos-baker&field.status_filter=published

### Downloading zcash-params

> Since v8.0 it's required to have `sapling-output.params` and `sapling-spend.params` files from

Reviewer: ```suggestion
Since v8.0 it's required to have the `sapling-output.params` and `sapling-spend.params` files from
```

the `zcash-params` within the `XDG_DATA_DIRS` to run the `tezos-node`.

> We'll need these params to be accessed by different
> users (because systemd services are usually run as a separate user), thus the suggested way of getting
> them via `https://raw.githubusercontent.com/zcash/zcash/master/zcutil/fetch-params.sh` script which will
> download them to the `HOME` directory of the current user is not convenient. Moreover, it will
> download large `sprout-groth16.params` which is unused at the runtime.

Reviewer: ```suggestion

We'll need these params to be accessible by different
users (because `systemd` services are usually run as a separate user), thus the suggested way of getting
them via [the zcash script](https://raw.githubusercontent.com/zcash/zcash/master/zcutil/fetch-params.sh), which will
download them to the `HOME` directory of the current user, is not convenient. Moreover, it will
download the large `sprout-groth16.params` which is unused at the runtime.
```

So, it's better to download them manually to the `/usr/local/share/zcash-params`, in order
to do that run the following commands:
```
sudo mkdir -p /usr/local/share/zcash-params
sudo wget https://gitlab.com/tezos/opam-repository/-/raw/v8.1/zcash-params/sapling-output.params -O /usr/local/share/zcash-params/sapling-output.params
sudo wget https://gitlab.com/tezos/opam-repository/-/raw/v8.1/zcash-params/sapling-spend.params -O /usr/local/share/zcash-params/sapling-spend.params

> ```

Reviewer: What do you think about making our own script just for this and putting it into `tezos-packaging/scripts/` instead?

It could be a copy of the one mentioned above (but with a different destination and without `groth16`) or perhaps a simpler new one.

Author: Apparently, it makes some sense

## Setting up the tezos-node

### Bootstrapping the node

> In order to run baker locally, you'll need fully-synced local `tezos-node`.

Reviewer: ```suggestion
In order to run a baker locally, you'll need a fully-synced local `tezos-node`.

```

The easiest way to set up a node running on the `mainnet` or on one of the
testnets (e.g. `delphinet` or `edonet`) is to use one of the predefined
`tezos-node-<network>.services` systemd services provided in the `tezos-node`
package. However, by default, these services will start to bootstrap the node

> from scratch, which will take significant amount of time. The fastest way to bootstrap
> the node is to import a snapshot. The snapshots can be downloaded from the https://snapshots.tulip.tools/#/ .

Reviewer: ```suggestion
from scratch, which will take significant amount of time.

The fastest way to bootstrap the node is to import a snapshot.
Snapshots can be downloaded from https://snapshots.tulip.tools/#/ .
```

`tezos-node-<network>.service` has `/var/lib/tezos/node-<network>` as a data directory

> and `http://localhost:8732` as a an RPC address by default.

Reviewer: ```suggestion
and `http://localhost:8732` as its RPC address by default.
```

These parameters can be changed in the service configuration, which is located in
`/lib/systemd/system/tezos-node-<network>.service`. Note, that if you've updated the `.service`
file, you should reload systemctl daemon via running `sudo systemctl daemon-reload`.

> All commands within the service are run under `tezos` user.

Reviewer: ```suggestion
All commands within the service are run under the `tezos` user.
```

> Download the snapshost for desired network and with desired history mode

Reviewer: ```suggestion
Download the snapshost for desired network and with the desired history mode
```

Reviewer: One more thing: it's a bit odd that the snapshots and were to find them is separated to this sentence by the part explaining the dir and parameter of the node.

Perhaps this section that's now in the middle can be moved above (just before the sentences that mention snapshots)? Or maybe it would be better to move it just below this?

Author: Yes, I think that the paragraph that starts from '`tezos-node-<network>.service` has `/var/lib/tezos/node-<network>` as a data directory' can be moved to the very top of the 'Setting up the tezos-node' part, right before the 'Bootstrapping the node'

(you can read more about `tezos-node` history modes [here](https://tezos.gitlab.io/user/history_modes.html#history-modes)).

> Snapshot import requires `tezos-node` data directory doesn't contain `context` and `store` folders.

Reviewer: ```suggestion
Snapshot import requires that the `tezos-node` data directory doesn't contain the `context` and `store` folders.
```

Remove them before the snapshot import.

In order to import the snapshot run the following command:
```
sudo -u tezos-node snapshot import --data-dir /var/lib/tezos/node-<network>/ <path to the snapshot file>
```

In case you're getting the error similar to
```
tezos-node: Error:
              Invalid block BLjutMj47caB
                Failed to validate the economic-protocol content of the block: Error:
                                                                                Invalid signature for block BLjutMj47caB. Expected: tz1Na5QB98cDA.
```

You should init/update config in the data directory to match the desired network
before importing the snapshot. To do the run the following:
```
sudo -u tezos tezos-node config init --data-dir /var/lib/tezos/node-<network> --network <network>
#or
sudo -u tezos tezos-node config update --data-dir /var/lib/tezos/node-<network> --network <network>

> ```

Reviewer: I suppose these aren't interchangeable, right?

Wouldn't the `init` one always be necessary the first time around and the `update` one after that?

Author: Yes, only one of these two commands will work. It depends on whether the config was initialized previously by the node startup script.

Apparently, the wording should be changed to something like ''To do the run one of the following commands depending on whether the config in the given data directory was initialized previously:"

### Starting the node

After the snapshot import, you can finally start the node by running
```
sudo systemctl start tezos-node-<network>.service
```

> Note that even after the snapshot import node can still be out of sync, it may require

Reviewer: ```suggestion
Note that even after the snapshot import the node can still be out of sync, it may require
```

some additional time to completely bootstrap.

To stop the node run:
```
sudo systemctl stop tezos-node-<network>.service
```

> You can check node logs via `journalctl`:

Reviewer: ```suggestion
You can check the node logs via `journalctl`:
```

```
journalctl -u tezos-node-<network>.service
```

## Setting up baker and endorser daemons

> Once you have fully synced local `tezos-node`, you can start baker and endorser daemons.

Reviewer: ```suggestion
Once you have fully synced your local `tezos-node`, you can start the baker and endorser daemons.
```

Data directories for baker and endorser daemons are defined in the
`/etc/default/tezos-baker-<proto>` and `/etc/default/tezos-endorser-<proto>`. Make
sure to point them at the same directory to have the same set of known keys, e.g.
point both `DATA_DIR`s to `/var/lib/tezos/baker`. You should also
update `NODE_DATA_DIR` in the `/etc/default/tezos-baker-<proto>` to point at the desired
node data directory, e.g. `/var/lib/tezos/node-<network>`.

Create the aformentioned daemons data directory:
```
sudo -u tezos mkdir -p /var/lib/tezos/baker
```

> Import you baker secret key to the data directory. There are two ways to import

Reviewer: ```suggestion
Import your baker secret key to the data directory. There are two ways to import
```

such key:

> 1) You know either unencrypted or password-encrypted secret key for your address. In order
> to import such key run:

Reviewer: ```suggestion
1) You know either the unencrypted or password-encrypted secret key for your address.

In order to import such key run:
```

```
sudo -u tezos tezos-client -d /var/lib/tezos/baker import secret key <alias> <secret-key>
```

> 2) Secret key is stored on ledger. Open Tezow Wallet app on your ledger and run the following

Reviewer: ```suggestion
2) The secret key is stored on a ledger.

Open the Tezoz Wallet app on your ledger and run the following
```

to import the key:
```
sudo -u tezos tezos-client -d /var/lib/tezos/baker import secret key <alias> <ledger-url>
```

> Apart from importing the key, you'll also need to set it up for baking. Open Tezos Baking app

Reviewer: From the rendered doc it isn't quite clear if what follows is needed in both cases or not.

In fact I'd suggest to divide this whole "Setting up baker and endorser daemons" section better into sub-steps(I suppose with `h3`/`###`:
- Set up daemons data directory
- Import you baker secret key to the data directory
- Start the daemons

on your ledger and run the following:
```
sudo -u tezos tezos-client -d /var/lib/tezos/baker setup ledger to bake for <alias>
```

> After importing the key, you'll need to update `BAKER_ACCOUNT` and `ENDORSER_ACCOUNT` in
> `/etc/default/tezos-baker-<proto>` and `/etc/default/tezos-endorser-<proto>` respectively with
> accordance to the alias of imported key.

Reviewer: ```suggestion
After importing the key, you'll need to update the `BAKER_ACCOUNT` and `ENDORSER_ACCOUNT` in
`/etc/default/tezos-baker-<proto>` and `/etc/default/tezos-endorser-<proto>` respectively, in
accordance to the alias of the imported key.
```

> Once the key is imported and the configs are updated, you can start baker and endorser daemons:

Reviewer: ```suggestion
Once the key is imported and the configs are updated, you can start the baker and endorser daemons:
```

```
sudo systemctl start tezos-baker-<proto>.service
sudo systemctl start tezos-endorser-<proto>.service
```

> Note that if you're baking with the ledger key, you should have Tezos Baking app open.

Reviewer: ```suggestion
Note that if you're baking with the ledger key, you should have the Tezos Baking app open.
```

<!--

>    - SPDX-FileCopyrightText: 2019 TQ Tezos <https://tqtezos.com/>

Reviewer: Especially considering that this is a fresh file, let's start updating the copyright here.
```suggestion
   - SPDX-FileCopyrightText: 2021 TQ Tezos <https://tqtezos.com/>
```

Author: :see_no_evil:

### Setting up node service

`tezos-node-<network>.service` has `/var/lib/tezos/node-<network>` as a data directory
and `http://localhost:8732` as its RPC address by default.

In case you want to use different data directory or RPC address,
you should update the service configuration. To edit the service configuration run
```
sudo systemctl edit --full tezos-node-<network>.service
```

### Bootstrapping the node

In order to run a baker locally, you'll need a fully-synced local `tezos-node`.

The fastest way to bootstrap the node is to import a snapshot.
Snapshots can be downloaded from the following websites:
* https://snapshots.tulip.tools/#/
* https://snapshots-tezos.giganode.io/
* https://xtz-shots.io/

All commands within the service are run under the `tezos` user.

Download the snapshot for desired network. We recommend to use rolling snapshot, this is
the smallest and the fastest mode that is sufficient for baking (you can read more about other
`tezos-node` history modes [here](https://tezos.gitlab.io/user/history_modes.html#history-modes)).

In order to import the snapshot run the following command:
```

> sudo -u tezos tezos-node snapshot import --data-dir /var/lib/tezos/node-<network>/ <path to the snapshot file>

Reviewer: Testing this on a fresh VM it seems there are permission problems here for the `tezos` user.

Can you check that's the case?

Reviewer: Additionally, while we fix this I think that in the next release we should add aliases for `tezos-node-<network>` that are just `tezos-node` with the correct `data-dir`.

This would provide a great UX as these aliases names would match the service names as well.

Reviewer: Ok so #148 is blocked, let's not wait for it for this article.

```
tezos-node: Error:
              Invalid block BLjutMj47caB
                Failed to validate the economic-protocol content of the block: Error:
                                                                                Invalid signature for block BLjutMj47caB. Expected: tz1Na5QB98cDA.
```

You should init/update config in the data directory to match the desired network
before importing the snapshot. To do that run one of the following commands
depending on whether the config in the given data directory was initialized previously:
```
sudo -u tezos tezos-node config init --data-dir /var/lib/tezos/node-<network> --network <network>
#or
sudo -u tezos tezos-node config update --data-dir /var/lib/tezos/node-<network> --network <network>
```

### Starting the node

After the snapshot import, you can finally start the node by running
```
sudo systemctl start tezos-node-<network>.service
```

Note that even after the snapshot import the node can still be out of sync, it may require
some additional time to completely bootstrap. In order to check whether the node is bootstrapped,
you can use `tezos-client`:
```
tezos-client --endpoint <node-rpc-address> bootstrapped

> ```

Reviewer: Actually even better, if the tezos client default config had the right values this could be just replaced with:
```
sudo -u tezos tezos-client bootstrapped
```

To stop the node run:
```
sudo systemctl stop tezos-node-<network>.service
```

> 

Reviewer: I would add another small suggestion here about enabling the service as well, like:
```
If you want the service start automatically at boot use `enable`:
<enable example here>
and to disable it from doing so, instead use:
<disable example here>
```

You can check the node logs via `journalctl`:
```
journalctl -u tezos-node-<network>.service
```

## Setting up baker and endorser daemons

### Setting up daemons data directory

Data directories for baker and endorser daemons are defined in the
`/etc/default/tezos-baker-<proto>` and `/etc/default/tezos-endorser-<proto>`.
By default, both this daemons have `/var/lib/tezos/client` set as a `DATA_DIR`.

> Additionally, you should specify `NODE_DATA_DIR` in the `/etc/default/tezos-baker-<proto>`

Reviewer: ```suggestion
Additionally, you need to specify `NODE_DATA_DIR` in the `/etc/default/tezos-baker-<proto>`
```
AFAIU this is a necessity at this point in the tutorial

to point at the desired node data directory, e.g. `/var/lib/tezos/node-<network>`.

### Importing baker key

Import your baker secret key to the data directory. There are two ways to import
such key:
1) You know either the unencrypted or password-encrypted secret key for your address.

In order to import such key run:
```
sudo -u tezos tezos-client -d /var/lib/tezos/client import secret key <alias> <secret-key>
```
2) The secret key is stored on a ledger.

Open the Tezoz Wallet app on your ledger and run the following
to import the key:
```

> sudo -u tezos tezos-client -d /var/lib/tezos/client import secret key <alias> <ledger-url>

Reviewer: This and the one just above also could be used without the `-d` argument if we change the default config

```
Apart from importing the key, you'll also need to set it up for baking. Open Tezos Baking app
on your ledger and run the following:
```
sudo -u tezos tezos-client -d /var/lib/tezos/client setup ledger to bake for <alias>
```

After importing the key, you'll need to update the `BAKER_ACCOUNT` and `ENDORSER_ACCOUNT` in
`/etc/default/tezos-baker-<proto>` and `/etc/default/tezos-endorser-<proto>` respectively, in
accordance to the alias of the imported key.

### Starting daemons

Once the key is imported and the configs are updated, you can start the baker and endorser daemons:
```
sudo systemctl start tezos-baker-<proto>.service
sudo systemctl start tezos-endorser-<proto>.service
```

> If the node isn't bootstrapped yet, baker and endorser daemons will wait for it to bootstrap.

Reviewer: ```suggestion
If the node isn't bootstrapped yet, the baker and endorser daemons will wait for it to bootstrap.
```

suffix in lowercase, e.g. list of available baker packages can be found
[here](https://launchpad.net/~serokell/+archive/ubuntu/tezos/+packages?field.name_filter=tezos-baker&field.status_filter=published).

To install them run the following commands:
```
# Add PPA with Tezos binaries
sudo add-apt-repository ppa:serokell/tezos
sudo apt-get update
# Install packages
sudo apt-get install tezos-client tezos-node tezos-baker-<proto> tezos-endorser-<proto>
```

Packages for `tezos-node`, `tezos-baker-<proto>` and `tezos-endorser-<proto>` provide
systemd units for running the corresponding binaries in the background.

### Downloading zcash-params

Since v8.0 it's required to have the `sapling-output.params` and `sapling-spend.params` files from
the `zcash-params` within the `XDG_DATA_DIRS` to run the `tezos-node`.

We'll need these params to be accessible by different
users (because `systemd` services are usually run as a separate user), thus the suggested way of getting
them via [the zcash script](https://raw.githubusercontent.com/zcash/zcash/master/zcutil/fetch-params.sh), which will
download them to the `HOME` directory of the current user, is not convenient. Moreover, it will

> download the large `sprout-groth16.params` which is unused at the runtime.

Reviewer: ```suggestion
download the large `sprout-groth16.params` which is unused at runtime.
```
I think here the "the" is unnecessary.

So, it's better to download them to the `/usr/local/share/zcash-params`, the easiest way to do that
is to run [`fetch-sapling-params.sh`](../scripts/fetch-sapling-params.sh) script:
```

> sudo ./fetch-salping-params.sh

Reviewer: ```suggestion
sudo ./fetch-sapling-params.sh
```
Also, I wonder if we should provide an easier way to grab this, this kinda presumes that the repo has been cloned or at least the file downloaded, perhaps it would be better to include the command here that download an executes the script.

On the same point I guess is the link just above, but I think we'll cross that bridge when we'll know where this is going to end up.

Author: :facepalm:

```

Sudo is required because the files are downloaded to the `/usr/local/share/zcash-params`.
This script also checks that hashsums of the downloaded files match the expected values.

## Setting up the tezos-node

The easiest way to set up a node running on the `mainnet` or on one of the
testnets (e.g. `delphinet` or `edonet`) is to use one of the predefined
`tezos-node-<network>.services` systemd services provided in the `tezos-node`

> package. However, by default, these services will start to bootstrap the node
> from scratch, which will take significant amount of time.

Reviewer: ```suggestion
package. 

However, by default, these services will start to bootstrap the node from scratch,
which will take a significant amount of time.
In order to avoid this, here we suggest bootstrapping from a snapshot instead.
```
Just to push towards "keep reading" and not "start waiting" :sweat_smile:

and `http://localhost:8732` as its RPC address by default.
These parameters can be changed in the service configuration, which is located in
`/lib/systemd/system/tezos-node-<network>.service`. Note, that if you've updated the `.service`
file, you should reload systemctl daemon via running `sudo systemctl daemon-reload`.

### Bootstrapping the node

In order to run a baker locally, you'll need a fully-synced local `tezos-node`.

The fastest way to bootstrap the node is to import a snapshot.
Snapshots can be downloaded from https://snapshots.tulip.tools/#/ .

All commands within the service are run under the `tezos` user.

Download the snapshost for desired network and with the desired history mode
(you can read more about `tezos-node` history modes [here](https://tezos.gitlab.io/user/history_modes.html#history-modes)).

Snapshot import requires that the `tezos-node` data directory doesn't contain the `context` and `store` folders.
Remove them before the snapshot import.

In order to import the snapshot run the following command:
```
sudo -u tezos-node snapshot import --data-dir /var/lib/tezos/node-<network>/ <path to the snapshot file>
```

> In case you're getting the error similar to

Reviewer: ```suggestion
In case you're getting an error similar to
```

sudo add-apt-repository ppa:serokell/tezos
sudo apt-get update
# Install packages
sudo apt-get install tezos-client tezos-node tezos-baker-<proto> tezos-endorser-<proto>
```

Packages for `tezos-node`, `tezos-baker-<proto>` and `tezos-endorser-<proto>` provide
systemd units for running the corresponding binaries in the background.

### Downloading zcash-params

Since v8.0 it's required to have the `sapling-output.params` and `sapling-spend.params` files from
the `zcash-params` within the `XDG_DATA_DIRS` to run the `tezos-node`.

We'll need these params to be accessible by different
users (because `systemd` services are usually run as a separate user), thus the suggested way of getting
them via [the zcash script](https://raw.githubusercontent.com/zcash/zcash/master/zcutil/fetch-params.sh), which will
download them to the `HOME` directory of the current user, is not convenient. Moreover, it will
download the large `sprout-groth16.params` which is unused at runtime.
So, it's better to download them to the `/usr/local/share/zcash-params`, the easiest way to do that
is to run [`fetch-sapling-params.sh`](../scripts/fetch-sapling-params.sh) script:
```
sudo ./fetch-sapling-params.sh
```

> 

Reviewer 2: is there anything that prevents sapling param files from being packaged into deb and installed as a dependency? these files are, in fact a hard runtime dependency for most tezos binaries, so if installing tezos-node does not also install sapling param the installation is broken/incomplete

Author: Yes, I think it makes sense to add one more `tezos-sapling-params` package and use it as a dependency for all other packages. Created #132

Sudo is required because the files are downloaded to the `/usr/local/share/zcash-params`.
This script also checks that hashsums of the downloaded files match the expected values.

## Setting up the tezos-node

The easiest way to set up a node running on the `mainnet` or on one of the
testnets (e.g. `delphinet` or `edonet`) is to use one of the predefined
`tezos-node-<network>.services` systemd services provided in the `tezos-node`
package. 

However, by default, these services will start to bootstrap the node from scratch,
which will take a significant amount of time.
In order to avoid this, here we suggest bootstrapping from a snapshot instead.

### Setting up node data directory

`tezos-node-<network>.service` has `/var/lib/tezos/node-<network>` as a data directory
and `http://localhost:8732` as its RPC address by default.
These parameters can be changed in the service configuration, which is located in
`/lib/systemd/system/tezos-node-<network>.service`. Note, that if you've updated the `.service`

> file, you should reload systemctl daemon via running `sudo systemctl daemon-reload`.

Reviewer 2: users should be cautioned against editing .service files directly since they are part of the package; it [seems that current recommended way](https://serverfault.com/questions/413397/how-to-set-environment-variable-in-systemd-service?answertab=votes#tab-top) is to use `systemctl edit`

Reviewer 2: it looks like this section doesn't actually set up anything, it's just information; would be good to explain why user may want/need to change the defaults

### Bootstrapping the node

In order to run a baker locally, you'll need a fully-synced local `tezos-node`.

The fastest way to bootstrap the node is to import a snapshot.

> Snapshots can be downloaded from https://snapshots.tulip.tools/#/ .

Reviewer 2: it would be good to give some alternatives here, https://snapshots-tezos.giganode.io/ and https://xtz-shots.io/ in particular are great

All commands within the service are run under the `tezos` user.

> Download the snapshost for desired network and with the desired history mode

Reviewer 2: typo: snapshost -> snapshots

(you can read more about `tezos-node` history modes [here](https://tezos.gitlab.io/user/history_modes.html#history-modes)).

> 

Reviewer 2: I think it would be better to make a concrete recommendation here that would work for most new users and save them time/effort by not requiring them to do a research, e.g. "download rolling snapshot" - that's smallest, fastest and is sufficient for baking. Tradeoffs, limitations and further reading should be separated into a note or a footnote so that they can be easily skipped if a user wants to get to the end of the happy path quickly

Snapshot import requires that the `tezos-node` data directory doesn't contain the `context` and `store` folders.

> Remove them before the snapshot import.

Reviewer 2: similarly, this is not something a user should encounter if the instructions to get to this point were correct and followed precisely, so should be a troubleshooting note/footnote rather than part of main narrative

In order to import the snapshot run the following command:
```

> sudo -u tezos-node snapshot import --data-dir /var/lib/tezos/node-<network>/ <path to the snapshot file>

Reviewer: ```suggestion
sudo -u tezos tezos-node snapshot import --data-dir /var/lib/tezos/node-<network>/ <path to the snapshot file>
```
missing the user :smile: 

Also, I'm trying following this on a fresh VM and it seems that there is no `tezos` user at this point, should it have been created by some of the previous steps (I doubt it, not much has happened), by one of the following ones, or should it be created manually?
If the latter is the case I think just above where we mention that all commands are run under `tezos` it should also say how to create the user.

Reviewer: There's another problem at this point, if one has never `start`ed the node service with `systemctl` before there in no `/var/lib/tezos` directory and this command will fail (with an `Invalid data directory` error which is not very helpful).

So I suggest that we either:
1. add the commands to create the data directory manually in the `Setting up node data directory` section
2. rearrange this part of the guide a bit and have the user `start` the node in that section and the `stop` it and `import` the snapshot after.

What do you think?

P.s. starting the `node` seems to not create the `tezos` user automatically, unless I've done something wrong

Reviewer 2: `/var/lib/tezos` should be created as part of .deb installation

Author: I assumed that both `tezos` user and `/var/lib/tezos` are both created during the package installation, but looks like it's not true :confused:. 
I Will check that additionally

Author: The aforementioned problems with user and directory creations should be fixed now

```

In case you're getting an error similar to
```
tezos-node: Error:
              Invalid block BLjutMj47caB
                Failed to validate the economic-protocol content of the block: Error:
                                                                                Invalid signature for block BLjutMj47caB. Expected: tz1Na5QB98cDA.
```

You should init/update config in the data directory to match the desired network
before importing the snapshot. To do that run one of the following commands
depending on whether the config in the given data directory was initialized previously:
```

> sudo -u tezos tezos-node config init --data-dir /var/lib/tezos/node-<network> --network <network>

Reviewer 2: is it possible to avoid having to specify `--data-dir` in all the commands? isn't it coming from tezos node config file? isn't there a default?

Author: AFAIU, in case when  `--data-dir` isn't presented it'll use `$HOME/.tezos-node` by default. It's possible to specify `--config-file` instead of `--data-dir`, but you'll still have to provide at least one of these options

Reviewer 2: these commands are to be executed as  `tezos` user, who's $HOME directory should be set to `/var/lib/tezos` and then defaults work just fine; also, a new user on a happy path/fast track to running node and baking on `mainnet` - which I think is the primary use case - shouldn't need to do any of this at all because `mainnet` is the default. To simplify things further (for the user, not necessarily for the packager 😈 ) `config init` should probably run at package installation time, so that user only ever needs to run `config update`, and that also becomes a note - explanation of how to set up for a testnet instead of mainnet - rather than the main story line.

Author: `config init` call is now a part of `tezos-node` package postinsallation

Reviewer: There is now a problem with this, which is that the `--network` option does not work for `edo2net`, which I think will be somewhat common considering it's the only other network presented aside from `mainnet`

#or
sudo -u tezos tezos-node config update --data-dir /var/lib/tezos/node-<network> --network <network>
```

### Starting the node

After the snapshot import, you can finally start the node by running
```
sudo systemctl start tezos-node-<network>.service
```

Note that even after the snapshot import the node can still be out of sync, it may require

> some additional time to completely bootstrap.

Reviewer 2: how does one check if node is fully synced, both initially and on-going?

To stop the node run:
```
sudo systemctl stop tezos-node-<network>.service
```

You can check the node logs via `journalctl`:
```
journalctl -u tezos-node-<network>.service
```

## Setting up baker and endorser daemons

Once you have fully synced your local `tezos-node`, you can start the baker and endorser daemons.

> 

Reviewer 2: do you actually have to wait for the node to fully sync or can you start baker and endorser and they will figure out when the node is ready themselves?

Author: You can indeed start both daemons, but they won't do anything unless the node is completely bootstrapped

Reviewer 2: _unless_ or _until_? if user can set everything up upfront and later just needs to check  that everything looks good once the node is synced is less mental effort than pausing the entire set up until node is bootstrapped and resuming it later, having to remember where were you in this complex workflow

Author: > unless or until?

:facepalm: 
of course, until

### Setting up daemons data directory

Data directories for baker and endorser daemons are defined in the
`/etc/default/tezos-baker-<proto>` and `/etc/default/tezos-endorser-<proto>`. Make
sure to point them at the same directory to have the same set of known keys, e.g.
point both `DATA_DIR`s to `/var/lib/tezos/baker`. You should also
update `NODE_DATA_DIR` in the `/etc/default/tezos-baker-<proto>` to point at the desired
node data directory, e.g. `/var/lib/tezos/node-<network>`.

Create the aformentioned daemons data directory:
```
sudo -u tezos mkdir -p /var/lib/tezos/baker

> ```

Reviewer 2: why does this require user action at all? the defaults should be correct out of the box, `/var/lib/tezos/baker` should be part of `.deb` package

Author: Agree, created #133

to point at the desired node data directory, e.g. `/var/lib/tezos/node-<network>`.

### Importing the baker key

Import your baker secret key to the data directory. There are multiple ways to import
the key:

1) You have faucet JSON file from https://faucet.tzalpha.net/.

In order to activate account run:
```
sudo -u tezos tezos-client activate account <alias> <path-to-downloaded-json>
```

2) You know either the unencrypted or password-encrypted secret key for your address.

In order to import such a key, run:
```
sudo -u tezos tezos-client import secret key <alias> <secret-key>
```
3) The secret key is stored on a ledger.

Open the Tezoz Wallet app on your ledger and run the following
to import the key:
```
sudo -u tezos tezos-client import secret key <alias> <ledger-url>
```
Apart from importing the key, you'll also need to set it up for baking. Open Tezos Baking app
on your ledger and run the following:
```
sudo -u tezos tezos-client setup ledger to bake for <alias>
```

### Registering the baker
Once the key is imported, you'll need to register your baker, in order to do that run the following
command:
```
sudo -u tezos tezos-client register key <alias> as delegate
```

> Check blockchain explorer (https://tzkt.io/ or https://tzstats.com/) to see the baker status and

Reviewer: ```suggestion
Check a blockchain explorer (e.g. https://tzkt.io/ or https://tzstats.com/) to see the baker status and
```

baking rights of your account.

### Updating baker and endorser daemons configuration
Apart from that, you'll need to update the `BAKER_ACCOUNT` and `ENDORSER_ACCOUNT` (in
`/etc/default/tezos-baker-<proto>` and `/etc/default/tezos-endorser-<proto>` respectively) in
accordance to the **alias** of the imported key.

### Starting daemons

Once the key is imported and the configs are updated, you can start the baker and endorser daemons:
```
sudo systemctl start tezos-baker-<proto>.service
sudo systemctl start tezos-endorser-<proto>.service
```

If the node isn't bootstrapped yet, the baker and endorser daemons will wait for it to bootstrap.

Note that if you're baking with the ledger key, you should have the Tezos Baking app open.

> Once the services were started, you can check their logs via `journalctl`:

Reviewer: ```suggestion
Once the services are started, you can check their logs via `journalctl`:
```
Or `have been`, I guess it would also be correct.
